controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/http"
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/flynn/flynn/Godeps/_workspace/src/gopkg.in/inconshreveable/log15.v2"
+	"github.com/flynn/flynn/controller/client"
+	ct "github.com/flynn/flynn/controller/types"
+	"github.com/flynn/flynn/controller/utils"
+	"github.com/flynn/flynn/discoverd/client"
+	"github.com/flynn/flynn/host/types"
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/cluster"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/random"
+	"github.com/flynn/flynn/pkg/shutdown"
+	"github.com/flynn/flynn/pkg/status"
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+const (
+	eventBufferSize    int           = 1000
+	maxJobAttempts     uint          = 30
+	jobAttemptInterval time.Duration = 500 * time.Millisecond
+)
+
+var logger = log15.New("component", "scheduler")
+
+func fnLogger(ctx ...interface{}) log15.Logger {
+	pc := make([]uintptr, 10) // at least 1 entry needed
+	runtime.Callers(2, pc)
+	f := runtime.FuncForPC(pc[0])
+	name := f.Name()
+	name = name[strings.LastIndex(name, ".")+1:]
+	params := []interface{}{"fn", name}
+	params = append(params, ctx...)
+	return logger.New(params...)
+}
```

Comment:
I wish Go had macros like `__func__`. I added this because grepping and searching for function names became arduous because of the string getting doubled up. If you think the overhead is to much, we can add the names back in, but I do think this results in much cleaner code.